src/index.js

```diff
@@ -3,247 +3,267 @@
+    } else {
+      // No extract condition is given. Extract comments that match commentsOpts instead of preserving them
+      condition.preserve = false;
+      condition.extract = commentsOpts;
+    }
+
+    // Ensure that both conditions are functions
+    ['preserve', 'extract'].forEach((key) => {
+      let regexStr;
+      let regex;
+      switch (typeof (condition[key])) {
+        case 'boolean':
+          condition[key] = condition[key] ? () => true : () => false;
+          break;
+        case 'function':
+          break;
+        case 'string':
+          if (condition[key] === 'all') {
+            condition[key] = () => true;
+            break;
+          }
+          if (condition[key] === 'some') {
+            condition[key] = (astNode, comment) => comment.type === 'comment2' && /@preserve|@license|@cc_on/i.test(comment.value);
+            break;
+          }
+          regexStr = condition[key];
+          condition[key] = (astNode, comment) => new RegExp(regexStr).test(comment.value);
+          break;
+        default:
+          regex = condition[key];
+          condition[key] = (astNode, comment) => (regex.test(comment.value));
+      }
+    });
+
+    // Redefine the comments function to extract and preserve
+    // comments according to the two conditions
+    return (astNode, comment) => {
+      if (condition.extract(astNode, comment)) {
+        extractedComments.push(
+          comment.type === 'comment2' ? `/*${comment.value}*/` : `//${comment.value}`,
+        );
+      }
+      return condition.preserve(astNode, comment);
+    };
+  }
 
+  apply(compiler) {
     const requestShortener = new RequestShortener(compiler.context);
-    compiler.plugin('compilation', (compilation) => {
-      if (options.sourceMap) {
-        compilation.plugin('build-module', (module) => {
+    compiler.plugin('compilation', (compilationArg) => {
+      const compilation = compilationArg;
+
+      if (this.options.sourceMap) {
+        compilation.plugin('build-module', (moduleArg) => {
           // to get detailed location info about errors
-          module.useSourceMap = true;
+          const moduleVar = moduleArg;
+          moduleVar.useSourceMap = true;
         });
       }
+
       compilation.plugin('optimize-chunk-assets', (chunks, callback) => {
-        const files = [];
-        // eslint-disable-next-line prefer-spread
-        chunks.forEach(chunk => files.push.apply(files, chunk.files));
-        // eslint-disable-next-line prefer-spread
-        files.push.apply(files, compilation.additionalChunkAssets);
-        const filteredFiles = files.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options));
         const uglifiedAssets = new WeakSet();
-        filteredFiles.forEach((file) => {
-          const oldWarnFunction = uglify.AST_Node.warn_function;
-          const warnings = [];
-          let sourceMap;
-          try {
+        chunks.reduce((acc, chunk) => acc.concat(chunk.files || []), [])
```

Comment:
I guess that's a bit overkill, it's an habit (bad ?) I took, since I saw people not use empty array but null or something like. 
Would this kind of defaulting have negative effects on performance or side effects ?